Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm

```diff
@@ -443,31 +445,47 @@
     return true;
 }
 
-bool WebExtensionAPITabs::parseSendMessageOptions(NSDictionary *options, std::optional<WebExtensionFrameIdentifier>& frameIdentifier, NSString *sourceKey, NSString **outExceptionString)
+bool WebExtensionAPITabs::parseSendMessageOptions(NSDictionary *options, WebExtensionMessageTargetParameters& targetParameters, NSString *sourceKey, NSString **outExceptionString)
 {
     static NSDictionary<NSString *, id> *types = @{
         frameIdKey: NSNumber.class,
+        documentIdKey: NSString.class,
     };
 
     if (!validateDictionary(options, sourceKey, nil, types, outExceptionString))
         return false;
 
-    if (NSNumber *frameID = objectForKey<NSNumber>(options, frameIdKey)) {
+    if (options[frameIdKey] && options[documentIdKey]) {
+        *outExceptionString = toErrorString(nil, sourceKey, @"it cannot specify both 'frameId' and 'documentId'");
+        return false;
+    }
+
+    if (NSNumber *frameID = options[frameIdKey]) {
         auto identifier = toWebExtensionFrameIdentifier(frameID.doubleValue);
         if (!isValid(identifier)) {
             *outExceptionString = toErrorString(nil, frameIdKey, @"it is not a frame identifier");
             return false;
         }
 
-        frameIdentifier = identifier.value();
+        targetParameters.frameIdentifier = identifier.value();
+    }
+
+    if (NSString *documentID = options[documentIdKey]) {
```

Comment:
Nit: A bit strange to have `documentID` and `documentId` in the same line, but I guess `documentId` is what the API is named and `documentID` is our style?